os/drivers/input/touchscreen.c

```diff
@@ -0,0 +1,371 @@
+ * Name: touch_open
+ ****************************************************************************/
+
+static int touch_open(FAR struct file *filep)
+{
+	FAR struct touchscreen_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	if (priv->crefs == 0) {
+		priv->ops->touch_enable(priv);
+	}
+	priv->crefs++;
+	DEBUGASSERT(priv->crefs > 0);
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_close
+ ****************************************************************************/
+
+static int touch_close(FAR struct file *filep)
+{
+	FAR struct touchscreen_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	DEBUGASSERT(priv->crefs > 0);
+	priv->crefs--;
+	if (priv->crefs == 0) {
+		priv->ops->touch_disable(priv);
+#ifndef CONFIG_DISABLE_POLL
+		/* Check if this file is registered in a list of waiters for polling.
+		* For example, when task A is blocked by calling poll and task B try to terminate task A,
+		* a pollfd of A remains in this list. If it is, it should be cleared.
+		*/
+		(void)touch_semtake(&priv->pollsem, false);
+		for (int i = 0; i < CONFIG_TOUCH_NPOLLWAITERS; i++) {
+			struct pollfd *fds = priv->fds[i];
+			if (fds && (FAR struct file *)fds->filep == filep) {
+				priv->fds[i] = NULL;
+			}
+		}
+		touch_semgive(&priv->pollsem);
+#endif
+	}
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_read
+ ****************************************************************************/
+
+static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	FAR struct touchscreen_s *priv;
+	size_t outlen;
+	int ret;
+
+	priv = filep->f_inode->i_private;
+
+	if (!priv || !buffer || buflen < 1) {
+		return -EINVAL;
+	}
+	/* Wait for semaphore to prevent concurrent reads */
+	touch_semtake(&priv->sem, false);
+
+	/* Read the touch data, only if screen has been touched or if we're waiting for touch up */
+	outlen = sizeof(struct touch_sample_s);
+	ret = priv->ops->touch_read(priv, buffer);
+	touch_semgive(&priv->sem);
+	return ret < 0 ? ret : outlen;
+}
+
+/****************************************************************************
+ * Name: touch_write
+ ****************************************************************************/
+
+static ssize_t touch_write(FAR struct file *filep, FAR const char *buffer, size_t buflen)
+{
+	//TODO
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_ioctl
+ *
+ * Description:
+ *   This IOCTL is used to configure the touch screen device.
+ *
+ ****************************************************************************/
+
+static int touch_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	return OK;
```

Comment:
I think there should be suspend mode or something to handle sleep mode.
It is depend on requirement of product, but I think ioctl should provide some interface to handle suspend & wakeup to prevent touch interrupt during sleep mode.

Reply:
Hi Mr. kwon,
I have added a ioctl support for sleep mode and wake up mode.
the interrupts will be disabled in case of sleep mode and interrupts will be enabled in case of wake up mode.

Reply:
yes, i think so, it could be needed, but this PR is opening long ago. Let's seperate PR about this.